src/main/java/com/projectkorra/core/ability/AbilityUser.java

```diff
@@ -0,0 +1,296 @@
+package com.projectkorra.core.ability;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+
+import com.projectkorra.core.ability.activation.Activation;
+import com.projectkorra.core.event.user.UserBindChangeEvent;
+import com.projectkorra.core.event.user.UserBindCopyEvent;
+import com.projectkorra.core.event.user.UserCooldownEndEvent;
+import com.projectkorra.core.event.user.UserCooldownStartEvent;
+import com.projectkorra.core.skill.SkillHolder;
+import com.projectkorra.core.util.Events;
+
+import org.bukkit.Location;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.util.Vector;
+
+/**
+ * Class for identifying something that can activate abilities
+ */
+public abstract class AbilityUser extends SkillHolder {
+	
+	private AbilityBinds binds = new AbilityBinds();
+	private Map<String, Cooldown> cooldowns = new HashMap<>();
+	private Map<Activation, SourceInstance> sources = new HashMap<>();
+	private LivingEntity entity;
+
+	public boolean immune = false;
+
+	public AbilityUser(LivingEntity entity) {
+		this.entity = entity;
+	}
+
+	public LivingEntity getEntity() {
+		return entity;
+	}
+	
+	/**
+	 * Binds the given ability to the specified slot
+	 * @param slot Where to bind the ability on the hotbar, slots range [0, 8]
+	 * @param ability {@link Ability} to bind at the slot
+	 * @return false if the slot is out of bounds or the ability is null
+	 */
+	public final boolean bindAbility(int slot, Ability ability) {
+		if (slot < 0 || slot > 8 || ability == null) {
+			return false;
+		}
+
+		setSlot(slot, ability);
+		return true;
+	}
+
+	/**
+	 * Binds the given abilities to the specified slots, such that elements in slots
+	 * correspond to elements at the same index in abilities
+	 * @param slots Array of slots to bind to
+	 * @param abilities Array of abilities to bind
+	 * @return empty boolean array if given arrays do not match length, otherwise a boolen array returning the value of
+	 * {@link AbilityUser#bindAbility(int, Ability)} for each pair <code>slots[i], abilities[i]</code> that is the same length as the given arrays
+	 */
+	public final boolean[] bindAbilities(int[] slots, Ability[] abilities) {
+		if (slots.length != abilities.length) {
+			return new boolean[0];
```

Comment:
why not return null instead of an empty boolean array?

Reply:
The idea is that it's returning which slots were successfully bound, so that would be an indicator that no slots were successfully bound, but null would probably work too

Reply:
I think the empty array is fine, it means we dont need to check for NPE.